File: app/src/main/java/com/challange/hilt/ui/main/MainFragment.java
package com.challange.hilt.ui.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.challange.hilt.R;
import com.challange.hilt.ui.models.Challenge;
import com.savvyapps.togglebuttonlayout.ToggleButtonLayout;

import org.jetbrains.annotations.NotNull;

import dagger.hilt.android.AndroidEntryPoint;
import kotlin.Pair;

import static androidx.core.os.BundleKt.bundleOf;

@AndroidEntryPoint
public class MainFragment extends Fragment implements ChallengeListListener {
    private MainAdapter adapter;
    private MainViewModel viewModel;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ((AppCompatActivity) getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);
        return inflater.inflate(R.layout.main_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ToggleButtonLayout toggleButton = view.findViewById(R.id.toggle_button_layout);
        toggleButton.setToggled(R.id.toggle_left, true);
        if (adapter == null) {
            adapter = new MainAdapter(this);
        }
        adapter.setListener(this);
        RecyclerView recyclerView = view.findViewById(R.id.recyclerView);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));

        viewModel.getUiData().observe(getViewLifecycleOwner(), challenges -> adapter.updateList(challenges));

    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.main_fragment_menu, menu);
    }

    @Override
    public void onClick(@NotNull Challenge item) {
        NavController navController = Navigation.findNavController(requireActivity(), R.id.nav_host_fragment);
        navController.navigate(R.id.detail_fragment_dest, bundleOf(new Pair<>("item", item)));
    }
}

File: app/src/main/java/com/challange/hilt/ui/main/MainFragment.kt
package com.challange.hilt.ui.main

import android.os.Bundle
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.Navigation.findNavController
import androidx.navigation.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.challange.hilt.R
import com.challange.hilt.ui.detail.DetailFragment
import com.challange.hilt.ui.models.Challenge
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.android.synthetic.main.main_fragment.*
import timber.log.Timber

@AndroidEntryPoint
class MainFragment : Fragment(), ChallengeListListener {

    private val viewModel by viewModels<MainViewModel>()
    private val mainAdapter by lazy { MainAdapter(this) }
    private val navController by lazy { requireActivity().findNavController(R.id.nav_host_fragment) }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        (requireActivity() as AppCompatActivity).supportActionBar?.setDisplayHomeAsUpEnabled(false)

        return inflater.inflate(R.layout.main_fragment, container, false)
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        //FIXME : auto selection for toggle, remove this line once material toggle group is used
        toggle_button_layout.setToggled(R.id.toggle_left, true)

        recyclerView.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(requireContext())
            adapter = mainAdapter
        }

        viewModel.uiData.observe(viewLifecycleOwner, Observer { data ->
            Timber.i("Challenges received with size : %s", data.size)
            mainAdapter.updateList(data)
        })
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) =
        inflater.inflate(R.menu.main_fragment_menu, menu)

    override fun onClick(item: Challenge) {
        navController.navigate(R.id.detail_fragment_dest, bundleOf("item" to item))
    }
}